vm/range_test.go

```diff
@@ -294,3 +294,169 @@ func TestInclude(t *testing.T) {
 		vm.checkCFP(t, i, 0)
 	}
 }
+
+func TestBsearch(t *testing.T) {
+	tests := []struct {
+		input    string
+		expected interface{}
+	}{
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(0..4).bsearch do |i|
+			ary[i] >= 0
+		end
+		`, 0},
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(0..4).bsearch do |i|
+			ary[i] >= 4
+		end
+		`, 1},
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(2..4).bsearch do |i|
+			ary[i] >= 4
+		end
+		`, 2},
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(4..4).bsearch do |i|
+			ary[i] >= 4
+		end
+		`, 4},
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(0..4).bsearch do |i|
+			ary[i] >= 6
+		end
+		`, 2},
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(0..4).bsearch do |i|
+			ary[i] >= 8
+		end
+		`, 3},
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(0..2).bsearch do |i|
+			ary[i] >= 8
+		end
+		`, nil},
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(0..4).bsearch do |i|
+			ary[i] >= 100
+		end
+		`, nil},
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(4..0).bsearch do |i|
+			ary[i] >= 4
+		end
+		`, nil},
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(-1..3).bsearch do |i|
+			ary[i] >= 4
+		end
+		`, nil},
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(1..-2).bsearch do |i|
+			ary[i] >= 4
+		end
+		`, nil},
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(-5..-2).bsearch do |i|
+			ary[i] >= 4
+		end
+		`, nil},
+		{`
+		ary = [0, 100, 100, 100, 200]
+		(0..4).bsearch do |i|
+			100 - ary[i]
+		end
+		`, 2},
+		{`
+		ary = [0, 100, 100, 100, 200]
+		(0..4).bsearch do |i|
+			200 - ary[i]
+		end
+		`, 4},
+		{`
+		ary = [0, 100, 100, 100, 200]
+		(0..4).bsearch do |i|
+			0 - ary[i]
+		end
+		`, 0},
+		{`
+		ary = [0, 100, 100, 100, 200]
+		(2..4).bsearch do |i|
+			100 - ary[i]
+		end
+		`, 3},
+		{`
+		ary = [0, 100, 100, 100, 200]
+		(2..4).bsearch do |i|
+			0 - ary[i]
+		end
+		`, nil},
+		{`
+		ary = [0, 100, 100, 100, 200]
+		(-1..4).bsearch do |i|
+			0 - ary[i]
+		end
+		`, nil},
+		{`
+		ary = [0, 100, 100, 100, 200]
+		(4..0).bsearch do |i|
+			0 - ary[i]
+		end
+		`, nil},
+		{`
+		ary = [0, 100, 100, 100, 200]
+		(2..-1).bsearch do |i|
+			0 - ary[i]
+		end
+		`, nil},
+		{`
+		ary = [0, 100, 100, 100, 200]
+		(-5..-1).bsearch do |i|
+			0 - ary[i]
+		end
+		`, nil},
+	}
+
+	for i, tt := range tests {
+		vm := initTestVM()
+		evaluated := vm.testEval(t, tt.input)
+		checkExpected(t, i, evaluated, tt.expected)
+	}
+}
+
+func TestBsearchFail(t *testing.T) {
+	vm := initTestVM()
+	testsFail := []struct {
+		input    string
+		expected *Error
+	}{
+		{`
+		ary = [0, 4, 7, 10, 12]
+		(0..4).bsearch do |i|
+			"Binary Search"
+		end
+		`, initErrorObject(TypeErrorClass, "Expect Integer or Boolean type. got=%T", vm.initStringObject("Binary Search"))},
+	}
```

Comment:
@st0012 Reformatted, however, I need to put the `vm := initTestVM()` expression in the beginning of the function testing because I need to use it for error testing